.gitlab-ci.yml

```diff
@@ -0,0 +1,27 @@
+stages:          # List of stages for jobs, and their order of execution
+  - build-docker
+  - test
+
+
+docker_build:
+  stage: build-docker
+  image: docker:20.10
+  script:
+    - DOCKER_HOST=tcp://localhost:2375 docker login -u $CI_REGISTRY_USER -p $CI_JOB_TOKEN $CI_REGISTRY
+    - docker build -t $CI_REGISTRY_IMAGE:$CI_BUILD_REF -f BaseDockerfile .
+    - docker push $CI_REGISTRY_IMAGE:$CI_BUILD_REF
```

Comment:
Does this step add a lot of time to CI runs? Is there a way to only rebuild less frequently than every commit, e.g. only if the CI Dockerfile has changed?

Reply:
Since this dockerfile is only the base environment it adds no time if the machine has already has the container layers built, which that is correct.

Reply:
Ah, got it; I guess we might want to revisit this if we move to autoscaling/ephemeral instances in the future, but makes sense for persistent instances

Reply:
A couple other questions/comments here:
- would this push a new image to ECR on every commit to a feature branch? This seems like it could create a lot of clutter
- I couldn't find documentation for `CI_BUILD_REF` [here](https://docs.gitlab.com/ee/ci/variables/index.html#predefined-cicd-variables). It seems like this might be deprecated, at least according to this [issue](https://gitlab.gwdg.de/genepaint/gp3/-/issues/122)?

Reply:
This will only push layers that have changed, which since the base stage is cached based on the `environment.yml` and `requirements.txt` should change relatively rarely. It does create a lot of cluttered tags, which we can set to clean up on a regular basis. This seemed like a reasonable trade off to rebuilding everything from scratch on every build, but the build only adds 10 minutes

Good catch, will change to `CI_COMMIT_REF_NAME` which should only create one tag per branch.